### PR TITLE
[SYCL][Doc] Add sycl complex to complex algorithms extension

### PR DESCRIPTION
**Pull request by AidanBeltonS**
This PR extends the complex algorithms extension to support `sycl::ext::oneapi::complex` and `marray<sycl::ext::oneapi::complex>`. Additionally it adds the `multiplies` operator as a valid binary operation for complex values when reducing and scanning across work items. This PR has a dependency upon #6550.

### CODE REVIEW

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_algorithms.asciidoc [hunk 1]
```diff
@@ -50,6 +52,13 @@ John Pennycook, Intel (john 'dot' pennycook 'at' intel 'dot' com)
 
 This extension is written against the SYCL 2020 specification, Revision 4.
 
+This extension builds on the `sycl::ext::oneapi::complex` and
+`marray<sycl::ext::oneapi::complex>` classes. Therefore this extension is
+dependent on link:sycl_ext_oneapi_complex.asciidoc[sycl_ext_oneapi_complex]
+and
+link:sycl_ext_oneapi_complex_marray.asciidoc[sycl_ext_oneapi_complex_marray].
```
Comment (Pennycook) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_algorithms.asciidoc:59]: Prior to this change, developers could use the `std::complex` features without requiring those other two extensions.  Does anything in our `std::complex` support now depend on these new extensions?  If not, I think we should be explicit that only those parts of this extension which use `sycl::complex` or `marray` are dependent.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_algorithms.asciidoc [hunk 2]
```diff
@@ -64,33 +73,42 @@ value to determine which of the extension's APIs the implementation supports.
...
 - `group_shift_right`
 - `permute_group_by_xor`
 - `select_from_group`
 
-Usage of `std::complex<double>` requires support for double precision,
-and specifically the `sycl::aspect::fp64` device aspect.
+Usage of `gencomplex` and `mgencomplex` with double value type requires support
+for double precision, and specifically the `sycl::aspect::fp64` device aspect.
+Similarly, usage of `gencomplex` and `mgencomplex` with half value type
+requires support for half-precision, with `sycl::aspect::fp16`.
```
Comment (Pennycook) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_algorithms.asciidoc:105]: Nit: This is a little confusing to me, because `gencomplex` and `mgencomplex` refer to all the value types.  "`gencomplex` with `half` value type" seems less precise than "`sycl::complex<sycl::half>` and `sycl::marray<sycl::complex<sycl::half>>>`".

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_algorithms.asciidoc [hunk 3]
```diff
@@ -64,33 +73,42 @@ value to determine which of the extension's APIs the implementation supports.
...
 
 == Extended Functions and Algorithms
 
-The following group functions and algorithms accept `std::complex<float>`
-and `std::complex<double>` arguments only when used with a `BinaryOperation`
-argument of `sycl::plus`:
+The following group functions and algorithms accept `gencomplex` arguments
+only when used with a `BinaryOperation` argument of `sycl::plus` and
+`sycl::multiplies`. These operations do not support `mgencomplex`:
```
Comment (Pennycook) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_algorithms.asciidoc:111]: This change seems unrelated to the `sycl::complex` and `sycl::marray` parts of things.  Can you say more about why you're lifting this restriction, and what SPIR-V instruction(s) you expect it to generate?

**Review by Pennycook — CHANGES_REQUESTED** [review]

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_algorithms.asciidoc (hunk 3 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_algorithms.asciidoc:114]: Yes this change is somewhat unrelated to `sycl::complex` and `sycl::marray<complex>`, this restriction is being lifted to allow for a more fleshed out version of this extension. The multiplies operation is a valid binary operation for reductions and scans so we would like to support it. We currently do not have a valid SPIR-V instruction for the multiplies operation, so we would be adding a new one to support this op.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_algorithms.asciidoc (hunk 2 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex_algorithms.asciidoc]: Agreed ya this is confusing. I have updated this bit to, hopefully, be much clearer.

**Review by Pennycook — APPROVED** [review]
Changes LGTM.  Thanks!
